### DISCUSSION TITLE
css.global should return a component

### DISCUSSION THREAD

**Discussion by fatfisz** [stitchesjs/stitches#285] [Ideas] [answered]
I just rewrote a small project to use Stitches and I love it! There are still a few rough edges, but I really like the ideas behind this tool. For now, there is one thing that stands out to me the most.

I was trying to figure out how to use `css.global` and I'm not sure if the current implementation is as useful as it could be. First let's take a look at prior art:

- In styled-components you get [a component that applies global styles](https://styled-components.com/docs/api#createglobalstyle) when it's rendered
- Emotion has [something similar](https://emotion.sh/docs/globals)
- Even styled-jsx, which is kind-of hijacking the `<style>` tag, [accepts a global prop](https://github.com/vercel/styled-jsx#global-styles) with a similar behavior

The common pattern that those solutions enable is to apply global styles conditionally; it's also clear when that happens - it's whenever those components render. In addition, when those components unmount, the styles disappear.

If I understood correctly, neither the `latest` version nor the `canary` version currently support that kind of a pattern. The styles are rendered ASAP in the first case, and when the function returned by `css.global` is called in the second one, which is only useful if you never want to change them.

Would you be open to tackling this?

**Comment by peduarte** [comment]
@fatfisz hey! thanks so raising this :) I've moved it to a _GH discussion_ so we can talk about and will then create an actionable issue if needed.

For context: I'll lay out the API differences between `0.0.2` (latest) and `0.0.3-canary.4`:

**latest**
In this API, you could just call this function at the top of the file. But we noticed that its styles were not being server-side rendered
```js
css.global({
  body: {
    margin: 0
  }
})
```

so in **canary** we changed it to:
```jsx
const globalStyles = css.global({
  body: {
    margin: 0
  }
})

export const App = () => {
  globalStyles()

  return <div />
}
```

> The common pattern that those solutions enable is to apply global styles conditionally; 

Never thought about this before. Why would global styled be conditionally applied? Doesnt that defeits the point of it being global?

>  In addition, when those components unmount, the styles disappear.

I don't understand why this is a good idea - for global styles

---

I'm curious to see what's your use case, and hopefully, we can understand it more and take that into account.

Thanks!

**Reply by fatfisz** [reply]
> Why would global styled be conditionally applied? Doesnt that defeits the point of it being global?

My favorite example is a modal window: when it's open, it's best to set `overflow: hidden` on `<body>` so that when you scroll inside the modal, the rest of the document is not affected. After the modal is closed, `overflow: hidden` should disappear again:

```jsx
const HideOverflow = css.global({
  overflow: 'hidden',
});

function Modal() {
  const [show, setShow] = useState(false);
  ...
  return (
    <>
      {show && <HideOverflow />}
      ... // modal stuff
    </>
  )
}
```

A second example would applying styles to elements outside of the React tree (it might sound contrived, but it's an actual case I encountered at work). There was a 3rd-party plugin that needed some global styles based on the data from the React part. So whenever the data changed, global styles would need to change too.

```jsx
const EditorStyles = css.global({
  variants: {
    light: {
      '.3rd-party-class': {
        color: 'magenta',
      },
    },
    dark: {
      '.3rd-party-class': {
        color: 'cyan',
      },
    },
  },
});

function EditorWrapper() {
  const root = useRef();
  useEffect(() => {
    const editor = initEditor(root.current);
    return () => editor.destroy();
  }, []);
  return (
    <>
      <div ref={root} />
      <EditorStyles theme={useThemeNameFromContext()} />
    </>
  );
}
```

Of course the actual code was different, but here I wanted to give an idea of what I expect `css.global` could work like.

---

So I'd sum it up like so: more flexibility is needed when global styles might change depending on the state of the app. Also sometimes we don't have a 100% React app and it would be great to be able to handle those scenarios too. In this case I'd argue that `variants` and the `css` prop would also work for a component returned by `css.global`.

**Reply by peduarte** [reply]
Thanks for that, very helpful insights <3

What API do you have in mind for this?

**Reply by fatfisz** [reply]
The one I used in my examples: `css.global` would take an object with the same style declaration as `styled` (variants included), and would return a component that would have props derived from variants. `as`, `className`, and other props don't really make sense in this scenario.

**Comment by peduarte** [comment, marked as answer]
Hi! we've now released Stitches 0.1.0 and we have decided to stick with this API for now. Thanks!
